[Intel MKL] Add FP32 fusion of MatMul and Relu.

Zantares opened the pull request:
This the 2nd part of Enable FP32 FusedMatMul(MatMul with BiasAdd) for MKL-DNN backend(https://github.com/tensorflow/tensorflow/pull/31782). It will fuse Relu/Relu6/Elu with FusedMatMul and rewrite it to MklFusedMatMul.


Modified:
- tensorflow/core/graph/mkl_layout_pass.cc
- tensorflow/core/kernels/mkl_fused_ops_test.cc
- tensorflow/core/kernels/mkl_matmul_op_fused.cc
- tensorflow/core/kernels/mkl_matmul_ops_common.h

Signed-off-by: Lu Teng teng.lu@intel.com

Zantares committed:
Add FP32 fusion of MatMul and Relu.

Zantares commented:
The previous PR(#31782) will fuse MatMul and BiasAdd into FusedMatMul, and rewrite it to MklFusedMatMul, but it will prevent 2 situations:
- Attribute transpose_a is false.
- FusedMatMul must has only one post op(default is BiasAdd, but it will be fused with BiasAdd and Relu if possible). 

The 2nd rule will impact performance of some models like NCF because FusedMatMul wasn't rewritten to MklFusedMatMul and went into Eigen path. This PR will optimize it to MKL-DNN and remove the 2nd limitation by fusing it with Relu/Relu6/Elu.

ashraf-bhuiyan commented:
@penpornk please let us know if we can do anything for the PR to merge. We are thinking to have this PR merged before TF 2.1 cutoff.

penpornk reviewed (changes requested):
Thank you for the PR and I'm so sorry for my delay! I have one minor question.

Zantares commented on tensorflow/core/kernels/mkl_fused_ops_test.cc:
```diff
@@ -618,87 +617,137 @@ template <typename T>
+      if (std::find(fused_ops.begin(), fused_ops.end(), "Elu") !=
+          fused_ops.end()) {
+        last_op = "with_elu";
+        next_op = ops::Relu(root.WithOpName(last_op), next_op);
```
Thank for your reviewing, Relu will have the same result with Elu when the activation is greater than 0. I checked the code and found that the test always generate positive float so that this error was hidden after testing. I'll correct it.

Zantares replied:
Fixed.

Zantares committed:
Fix Elu issue for MKL fusion.

Zantares commented:
Hi @penpornk ,thanks for your suggestion, we've found an internal issue related to Elu in MKL back-end:
We used `SetRandom()` in UT to test the fusion ops, but it only generate positive value so that `Elu` become the same as `Relu`, then we weren't aware the typo error in UT. Besides, I found that we set wrong `alpha` for `Elu` and got wrong result after I corrected UT. `Alpha` in `Elu` should be `1` but not `0`: https://github.com/intel/mkl-dnn/blob/master/src/common/math_utils.hpp#L189

Now  I have fixed it with this PR, please take a look at the new changes, thanks!

Zantares commented:
Close this PR because migrated to GitLab: https://gitlab.devtools.intel.com/TensorFlow/Direct-Optimization/private-tensorflow/merge_requests/30

Zantares committed:
Separate Conv correction from this PR.

Zantares commented:
I made a new change after an internal discussion - separate the Conv correction part from this PR and will fix it in another PR. It will make bug fix and feature implementation more clear.

penpornk commented on tensorflow/core/kernels/mkl_fused_ops_test.cc:
```diff
@@ -242,7 +242,7 @@ class MklFusedConv2DOpTest : public OpsTestBase {
         fused_ops.end()) {
       last_op = "with_elu";
-      next_op = ops::Elu(root.WithOpName(last_op), next_op);
+      next_op = ops::Relu(root.WithOpName(last_op), next_op);
```
Could you please add a TODO in the code, saying that this is intentional and will be fixed later? Please use `TODO(<your_username>):` or `TODO(intel-tf):`. Thank you!

Zantares replied:
There're 2 Elu issues, I've fixed one related MatMul in this PR, and added comment for another related to Conv. Please take a look, thanks!

penpornk reviewed (changes requested):
Thank you very much for the careful investigation and explanation!

Zantares committed:
Add comment for known Conv and Elu fusion.

penpornk reviewed (approved):
Thank you very much!
